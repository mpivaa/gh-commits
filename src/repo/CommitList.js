// @flow
import * as React from 'react';
import styled from 'styled-components';
import fuzzysearch from 'fuzzysearch';
import { spacing, color } from '../theme';
import { Normal, Avatar, TextField } from '../ui';
import type { Commit } from './types';

let CommitListStyle = styled.ul`
  flex: 1;
  overflow-y: auto;
`;

let CommitItem = styled.li`
  display: flex;
  align-items: center;
  padding: ${spacing('xxs')} 0;
`;

let Message = styled(Normal)`
  flex: 1;
  color: ${color('grey')};
  margin-right: ${spacing('sm')};
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
`;

let CommitID = styled(Normal)`
  flex: 0;
  font-weight: bold;
  margin-right: ${spacing('sm')};
`;

let Author = styled.div`
  display: flex;
  align-items: center;
`;

let AuthorName = styled(Normal)`
  width: 100px;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
  margin-left: ${spacing('sm')};
`;

type Props = {
  commits: Commit[]
};

type State = {
  search: string
};

class CommitList extends React.Component<Props, State> {
  state = { search: '' };

  render() {
    let { commits } = this.props;
    let { search } = this.state;

    // Github API doesn't provide a way to filter all commits
    let filteredCommits = commits.filter(commit =>
      fuzzysearch(search, commit.message)
    );

    return (
      <CommitListStyle>
        <TextField
          type="text"
          value={search}
          placeholder="Search..."
          onChange={e => this.setState({ search: e.target.value })}
        />
        {filteredCommits.map(commit => (
          <CommitItem key={commit.abbreviatedOid}>
            <CommitID>{commit.abbreviatedOid}</CommitID>
            <Message>{commit.message}</Message>
            <Author>
              <Avatar src={commit.author.avatarUrl} size={24} />
              <AuthorName>{commit.author.name}</AuthorName>
            </Author>
          </CommitItem>
        ))}
      </CommitListStyle>
    );
  }
}

export { CommitList };
