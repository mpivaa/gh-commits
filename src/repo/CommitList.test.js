import React from 'react';
import renderer from 'react-test-renderer';
import { StaticRouter } from 'react-router-dom';
import { CommitList } from './CommitList';

it('renders a list of commits', () => {
  let tree = renderer
    .create(
      <StaticRouter location="/somelogin/somerepo" context={{}}>
        <CommitList
          commits={[
            {
              abbreviatedOid: 'abc',
              message: 'Some message',
              author: {
                name: 'Some author',
                avatarUrl: '...'
              }
            }
          ]}
        />
      </StaticRouter>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
