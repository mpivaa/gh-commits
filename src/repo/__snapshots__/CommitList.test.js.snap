// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a list of commits 1`] = `
<ul
  className="sc-gzVnrw fTEsWS"
>
  <input
    className="sc-bZQynM gqaLFO"
    onChange={[Function]}
    placeholder="Search..."
    type="text"
    value=""
  />
  <li
    className="sc-htoDjs elnIVe"
  >
    <div
      className="sc-iwsKbI RAFUZ"
    >
      abc
    </div>
    <div
      className="sc-dnqmqq cdFNjO"
    >
      Some message
    </div>
    <div
      className="sc-gZMcBi hhbgDr"
    >
      <img
        alt="Avatar"
        className="sc-EHOje eKiJYG"
        size={24}
        src="..."
      />
      <div
        className="sc-gqjmRU fMOkYO"
      >
        Some author
      </div>
    </div>
  </li>
</ul>
`;
